File: src/main/java/no/fintlabs/AcosInstanceMapper.java
package no.fintlabs;

import no.fintlabs.gateway.instance.InstanceMapper;
import no.fintlabs.gateway.instance.model.File;
import no.fintlabs.gateway.instance.model.instance.InstanceObject;
import no.fintlabs.model.acos.AcosDocument;
import no.fintlabs.model.acos.AcosInstance;
import no.fintlabs.model.acos.AcosInstanceElement;
import org.springframework.http.MediaType;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;
import reactor.util.function.Tuple2;

import java.util.*;
import java.util.function.Function;
import java.util.stream.Collectors;

@Service
public class AcosInstanceMapper implements InstanceMapper<AcosInstance> {

    @Override
    public Mono<InstanceObject> map(
            Long sourceApplicationId,
            AcosInstance acosInstance,
            Function<File, Mono<UUID>> persistFile
    ) {
        return Mono.zip(
                        mapPdfFileToFileId(persistFile, sourceApplicationId, acosInstance),
                        mapDocumentsToInstanceObjects(persistFile, sourceApplicationId, acosInstance.getMetadata().getInstanceId(), acosInstance.getDocuments())
                )
                .map((Tuple2<UUID, List<InstanceObject>> formPdfFileIdAndDocumentInstanceElement) -> InstanceObject
                        .builder()
                        .valuePerKey(toValuePerKey(
                                acosInstance.getElements(),
                                formPdfFileIdAndDocumentInstanceElement.getT1()
                        ))
                        .objectCollectionPerKey(Map.of(
                                "vedlegg", formPdfFileIdAndDocumentInstanceElement.getT2()
                        ))
                        .build());
    }

    private Map<String, String> toValuePerKey(
            Collection<AcosInstanceElement> acosInstanceElements,
            UUID formPdfFileid
    ) {
        Map<String, String> valuePerKey = acosInstanceElements
                .stream()
                .collect(Collectors.toMap(
                        acosInstanceElement -> "skjema." + acosInstanceElement.getId(),
                        acosInstanceElement -> Optional.ofNullable(acosInstanceElement.getValue()).orElse("")
                ));

        valuePerKey.put("skjemaPdf", formPdfFileid.toString());

        return valuePerKey;
    }

    private Mono<UUID> mapPdfFileToFileId(
            Function<File, Mono<UUID>> persistFile,
            Long sourceApplicationId,
            AcosInstance acosInstance
    ) {
        return persistFile.apply(
                File
                        .builder()
                        .name("skjemaPdf")
                        .type(MediaType.APPLICATION_PDF)
                        .sourceApplicationId(sourceApplicationId)
                        .sourceApplicationInstanceId(acosInstance.getMetadata().getInstanceId())
                        .encoding("UTF-8")
                        .base64Contents(acosInstance.getFormPdfBase64())
                        .build()
        );
    }

    private Mono<List<InstanceObject>> mapDocumentsToInstanceObjects(
            Function<File, Mono<UUID>> persistFile,
            Long sourceApplicationId,
            String sourceApplicationInstanceId,
            Collection<AcosDocument> acosDocuments
    ) {
        return Flux.fromIterable(acosDocuments)
                .flatMap(acosDocument -> mapDocumentToInstanceObject(
                        persistFile,
                        sourceApplicationId,
                        sourceApplicationInstanceId,
                        acosDocument
                ))
                .collectList();
    }

    private Mono<InstanceObject> mapDocumentToInstanceObject(
            Function<File, Mono<UUID>> persistFile,
            Long sourceApplicationId,
            String sourceApplicationInstanceId,
            AcosDocument acosDocument
    ) {
        File file = toFile(sourceApplicationId, sourceApplicationInstanceId, acosDocument);
        return persistFile.apply(file)
                .map(fileId -> toInstanceObject(acosDocument, fileId));
    }

    private File toFile(
            Long sourceApplicationId,
            String sourceApplicationInstanceId,
            AcosDocument acosDocument
    ) {
        return File
                .builder()
                .name(acosDocument.getName())
                .sourceApplicationId(sourceApplicationId)
                .sourceApplicationInstanceId(sourceApplicationInstanceId)
                .type(acosDocument.getType())
                .encoding(acosDocument.getEncoding())
                .base64Contents(acosDocument.getBase64())
                .build();
    }

    private InstanceObject toInstanceObject(AcosDocument acosDocument, UUID fileId) {
        return InstanceObject
                .builder()
                .valuePerKey(Map.of(
                        "navn", acosDocument.getName(),
                        "type", acosDocument.getType().toString(),
                        "enkoding", Optional.ofNullable(acosDocument.getEncoding()).orElse(""),
                        "fil", fileId.toString()
                ))
                .build();
    }

}

File: src/main/kotlin/no/novari/flyt/acos/instance/gateway/AcosInstanceMapper.kt
package no.novari.flyt.acos.instance.gateway

import no.novari.flyt.acos.instance.gateway.model.acos.AcosDocument
import no.novari.flyt.acos.instance.gateway.model.acos.AcosInstance
import no.novari.flyt.acos.instance.gateway.model.acos.AcosInstanceElement
import no.novari.flyt.gateway.instance.InstanceMapper
import no.novari.flyt.gateway.instance.model.File
import no.novari.flyt.gateway.instance.model.instance.InstanceObject
import org.springframework.http.MediaType
import org.springframework.stereotype.Service
import java.util.UUID

@Service
class AcosInstanceMapper : InstanceMapper<AcosInstance> {
    override fun map(
        sourceApplicationId: Long,
        incomingInstance: AcosInstance,
        persistFile: (File) -> UUID,
    ): InstanceObject {
        val formPdfFileId = mapPdfFileToFileId(persistFile, sourceApplicationId, incomingInstance)
        val documentInstanceObjects =
            mapDocumentsToInstanceObjects(
                persistFile = persistFile,
                sourceApplicationId = sourceApplicationId,
                sourceApplicationInstanceId = incomingInstance.metadata.instanceId,
                acosDocuments = incomingInstance.documents,
            )

        return InstanceObject(
            valuePerKey = toValuePerKey(incomingInstance.elements, formPdfFileId),
            objectCollectionPerKey = mutableMapOf("vedlegg" to documentInstanceObjects),
        )
    }

    private fun toValuePerKey(
        acosInstanceElements: Collection<AcosInstanceElement>,
        formPdfFileId: UUID,
    ): Map<String, String> {
        val valuePerKey =
            acosInstanceElements
                .groupBy(AcosInstanceElement::id)
                .mapKeys { "skjema.${it.key}" }
                .mapValues { (_, elements) ->
                    elements.joinToString(separator = "\n") { it.value.orEmpty() }
                }.toMutableMap()

        valuePerKey["skjemaPdf"] = formPdfFileId.toString()

        return valuePerKey
    }

    private fun mapPdfFileToFileId(
        persistFile: (File) -> UUID,
        sourceApplicationId: Long,
        acosInstance: AcosInstance,
    ): UUID {
        return persistFile(
            File(
                name = "skjemaPdf",
                sourceApplicationId = sourceApplicationId,
                sourceApplicationInstanceId = acosInstance.metadata.instanceId,
                type = MediaType.APPLICATION_PDF,
                encoding = "UTF-8",
                base64Contents = acosInstance.formPdfBase64,
            ),
        )
    }

    private fun mapDocumentsToInstanceObjects(
        persistFile: (File) -> UUID,
        sourceApplicationId: Long,
        sourceApplicationInstanceId: String,
        acosDocuments: Collection<AcosDocument>,
    ): List<InstanceObject> {
        return acosDocuments.map { acosDocument ->
            mapDocumentToInstanceObject(
                persistFile = persistFile,
                sourceApplicationId = sourceApplicationId,
                sourceApplicationInstanceId = sourceApplicationInstanceId,
                acosDocument = acosDocument,
            )
        }
    }

    private fun mapDocumentToInstanceObject(
        persistFile: (File) -> UUID,
        sourceApplicationId: Long,
        sourceApplicationInstanceId: String,
        acosDocument: AcosDocument,
    ): InstanceObject {
        val file = toFile(sourceApplicationId, sourceApplicationInstanceId, acosDocument)
        val fileId = persistFile(file)
        return toInstanceObject(acosDocument, fileId)
    }

    private fun toFile(
        sourceApplicationId: Long,
        sourceApplicationInstanceId: String,
        acosDocument: AcosDocument,
    ): File {
        return File(
            name = acosDocument.name,
            sourceApplicationId = sourceApplicationId,
            sourceApplicationInstanceId = sourceApplicationInstanceId,
            type = acosDocument.type,
            encoding = acosDocument.encoding.orEmpty(),
            base64Contents = acosDocument.base64,
        )
    }

    private fun toInstanceObject(
        acosDocument: AcosDocument,
        fileId: UUID,
    ): InstanceObject {
        return InstanceObject(
            valuePerKey =
                mapOf(
                    "navn" to acosDocument.name,
                    "type" to acosDocument.type.toString(),
                    "enkoding" to acosDocument.encoding.orEmpty(),
                    "fil" to fileId.toString(),
                ),
        )
    }
}
